cirq/sim/op_by_op_simulator.py

```diff
@@ -0,0 +1,289 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A simulator that uses numpy's einsum for sparse matrix operations."""
+import abc
+import collections
+from typing import (
+    Dict,
+    Iterator,
+    List,
+    TYPE_CHECKING,
+    Tuple,
+    cast,
+    Set,
+    TypeVar,
+    Generic,
+)
+
+import numpy as np
+
+from cirq import circuits, ops, protocols, study, devices
+from cirq.sim import (
+    simulator,
+)
+from cirq.sim.abstract_state import AbstractState
+from cirq.sim.simulator import (
+    check_all_resolved,
+    SimulationTrialResult,
+    StepResult,
+    SimulatesIntermediateState,
+)
+
+if TYPE_CHECKING:
+    import cirq
+
+
+TState = TypeVar('TState', bound=AbstractState)
+
+
+class StateFactory(Generic[TState], metaclass=abc.ABCMeta):
```

Comment:
Please add class and method docstrings in this file to clarify how to implement these classes.